Extensions/retags.js

```diff
@@ -19,111 +19,168 @@ XKit.extensions.retags = {
 			// Activity
-			} else if ($t.hasClass('ui_note')) {
-				if ($t.find('.part_response').length) {
-					$t.addClass('is_response');
+			} else if ($element.hasClass('ui_note')) {
+				if ($element.find('.part_response').length) {
+					$element.addClass('is_response');
 				}
-				cls = 'is_retags';
-				$c = $t.find('.stage');
-				url = $c.find('.part_glass').attr('href');
+				retagClass = 'is_retags';
+				$target = $element;
+				url = $target.find('.part_glass').attr('href');
 			// dashboard
-			} else if ($t.hasClass('notification') && !XKit.browser().mobile) {
-				$c = $t.find('.notification_sentence');
-				url = $c.find('.notification_target').attr('href');
+			} else if ($element.hasClass('notification') && !XKit.browser().mobile) {
+				$target = $element.find('.notification_sentence');
+				url = $target.find('.notification_target').attr('href');
 			// mobile
-			} else if ($t.hasClass('notification') && XKit.browser().mobile) {
-				cls = 'is_retags';
-				$c = $t.find('.notification-wrapper');
-				url = $c.find('a').not('.notification-username').attr('href');
-			}
-			//we don't need to put tags on a reply, but we also don't need to hide it
-			if ($t.hasClass('is_reply') || $t.hasClass('is_answer') || $t.hasClass('type_6')) {
-				return;
+			} else if ($element.hasClass('notification') && XKit.browser().mobile) {
+				retagClass = 'is_retags';
+				$target = $element.find('.notification-wrapper');
+				url = $target.find('a').not('.notification-username').attr('href');
 			}
 			if (url) {
 				url = url.split('/');
-				var host = url[2], id = url[4], cache = 'post_'+id;
+				var host = url[2], id = url[4], cache = 'post_' + id;
 				if (XKit.storage.get('retags', cache, null) !== null) {
-					XKit.extensions.retags.append($t, cls, $c, XKit.storage.get('retags', cache));
+					var retags = XKit.extensions.retags.build_tag_collection($element, retagClass, XKit.storage.get('retags', cache));
+					$target.append(retags);
 				} else {
-					XKit.extensions.retags.request($t,cls,$c,cache,'https://api.tumblr.com/v2/blog/'+host+'/posts/info?id='+id+'&api_key='+XKit.extensions.retags.api_key);
+					var requestUrl = 'https://api.tumblr.com/v2/blog/' + host + '/posts/info?id=' + id + '&api_key=' + XKit.extensions.retags.api_key;
+					XKit.extensions.retags.request($element, retagClass, cache, requestUrl)
+						.then(function(retags) {
+							$target.append(retags);
+						});
 				}
 			}
 		});
 	},
 
-	request: function($t,cls,$c,cache,url) {
-		GM_xmlhttpRequest({
-			method: 'GET',
-			url: url,
-			onload: function(data){
-				var tags = JSON.parse(data.responseText).response.posts[0].tags;
-				if (XKit.extensions.retags.count_cached_posts() > XKit.extensions.retags.POST_CACHE_CLEAR_THRESHOLD || XKit.storage.quota('retags') < 100) {
-					XKit.extensions.retags.clear_old_posts();
+	request: function($element, retagClass, cache, url) {
+		return new Promise(function(resolve, reject) {
```

Comment:
is there a reason you can't just use $.get()? that returns a promise, so you don't have to mix the onload/promise APIs